core/src/main/java/com/google/zxing/qrcode/encoder/MinimalEncoder.java

```diff
@@ -0,0 +1,778 @@
+      case BYTE: return true; //any character can be encoded as byte(s). Up to the caller to manage splitting into multiple bytes when String.getBytes(Charset) return more than one byte.
+      default:
+        return false;
+    }
+  }
+
+  static int getCompactedOrdinal(Mode mode) {
+    if (mode == null) {
+      return 0;
+    }
+    switch (mode) {
+      case KANJI: return 0;
+      case ALPHANUMERIC: return 1;
+      case NUMERIC: return 2;
+      case ECI:
+      case BYTE: return 3;
+      default:
+        assert false;
+        return -1;
+    }
+  }
+
+  static ResultNode smallest(ResultNode[] results) {
+    ResultNode smallestResult = null;
+    for (int i = 0; i < results.length; i++) {
+      if (smallestResult == null || (results[i] != null && results[i].getSize() < smallestResult.getSize())) {
+        smallestResult = results[i];
+      }
+    }
+    return smallestResult;
+  }
+
+  static ResultNode smallest(Vector<ResultNode> results) {
+    ResultNode smallestResult = null;
+    for (int i = 0; i < results.size(); i++) {
+      if (smallestResult == null || (results.get(i) != null && results.get(i).getSize() < smallestResult.getSize())) {
+        smallestResult = results.get(i);
+      }
+    }
+    return smallestResult;
+  }
+
+  ResultNode postProcess(ResultNode result) {
+    if (isGS1) {
+      if (result.mode != Mode.ECI) {
+        ResultNode current = result.next;
+        while (current != null && current.mode != Mode.ECI) {
+          current = current.next;
+        }
+        if (current != null) { // there is an ECI somewhere
+          //prepend a default character set ECI
+          result = new ResultNode(Mode.ECI,result.version,true,0,0,result);
+        }
+      }
+      if (result.mode != Mode.ECI) {
+        //prepend a FNC1_FIRST_POSITION
+        result = new ResultNode(Mode.FNC1_FIRST_POSITION,result.version,true,0,0,result);
+      } else {
+        //insert a FNC1_FIRST_POSITION after the ECI
+        result.next = new ResultNode(Mode.FNC1_FIRST_POSITION,result.version,true,0,0,result.next);
+      }
+    }
+    ResultNode current = result;
+    while (current.next != null) {
+      current = current.next;
+    }
+    //Add TERMINATOR according to "8.4.8 Terminator"
+    current.next = new ResultNode(Mode.TERMINATOR,result.version,true,stringToEncode.length(),result.charsetEncoderIndex,null);
+    return result;
+  }
+    
+  void addEdge(Vector<ResultNode>[][][] vertices,ResultNode rn,ResultNode previous) {
+    int vertexIndex = rn.position + (rn.mode == Mode.ECI ? 1 : getEncodingGranularity(rn.mode));
+    if (vertices[vertexIndex][rn.charsetEncoderIndex][getCompactedOrdinal(rn.mode)] == null) {
+      vertices[vertexIndex][rn.charsetEncoderIndex][getCompactedOrdinal(rn.mode)] = new Vector<ResultNode>();
+    }
+    vertices[vertexIndex][rn.charsetEncoderIndex][getCompactedOrdinal(rn.mode)].add(rn);
+    rn = rn.getHead();
+    rn.setPrevious(previous);
+
+    if (DEBUG) {
```

Comment:
I generally have not put logging in the core library - I'd just remove this

Reply:
Are you sure? I find debugging the algorithm without it, difficult.

Reply:
How about commenting it?

Reply:
OK, comment it, or remove before the final merge